sdk/keyvault/Azure.Security.KeyVault.Keys/Readme.md

```diff
@@ -0,0 +1,282 @@
+
+## Key concepts
+### Keys
+Azure Key Vault supports multiple key types and algorithms, and enables the use of Hardware Security Modules (HSM) for high value keys.
+
+### Key Client:
+An asynchronous and synchronous KeyClient client exists in the SDK allowing for selection of a client based on an application's use case. Once you've initialized a KeyClient, you can interact with the primary resource types in Key Vault.
+
+## Examples
+The Azure.Security.KeyVault.Keys package supports synchronous and asynchronous APIs.
+
+The following section provides several code snippets using the [above created](#create-key-client) `client`, covering some of the most common Azure Key Vault Key service related tasks:
+
+### Sync examples
+* [Create a Key](#create-a-key)
+* [Retrieve a Key](#retrieve-a-key)
+* [Update an existing Key](#update-an-existing-key)
+* [Delete a Key](#delete-a-key)
+* [List Keys](#list-keys)
+
+### Async examples
+* [Create a Key](#async-create-a-key)
+
+### Create a Key
+Create a Key to be stored in the Azure Key Vault. If a key with the same name already exists, then a new version of the key is created.
+
+```c#
+// Create a key of any type
+Key key = client.CreateKey("key-name", KeyType.EllipticCurve);
+
+Console.WriteLine(key.Name);
+Console.WriteLine(key.KeyMaterial.KeyType);
+
+// Create a software RSA key
+var rsaCreateKey = new RsaKeyCreateOptions("rsa-key-name", hsm: false);
+Key rsaKey = Client.CreateRsaKey(rsaCreateKey);
+
+Console.WriteLine(rsaKey.Name);
+Console.WriteLine(rsaKey.KeyMaterial.KeyType);
+
+// Create a hardware Elliptic Curve key
+var echsmkey = new EcKeyCreateOptions("ec-key-name", hsm: true);
+Key ecKey = Client.CreateEcKey(echsmkey);
+
+Console.WriteLine(ecKey.Name);
+Console.WriteLine(ecKey.KeyMaterial.KeyType);
+```
+
+### Retrieve a Key
+`GetKey` retrieves a key previously stored in the Key Vault.
+
+```c#
+Key key = client.GetKey("key-name");
+
+Console.WriteLine(key.Name);
+Console.WriteLine(key.KeyMaterial.KeyType);
+```
+
+### Update an existing Key
+`UpdateKey` updates a key previously stored in the Key Vault.
+
+```c#
+Key key = client.CreateKey("key-name", KeyType.EllipticCurve);
+
+// You can specify additional application-specific metadata in the form of tags.
+key.Tags["foo"] = "updated tag";
+
+KeyBase updatedKey = client.UpdateKey(key);
+
+Console.WriteLine(updatedKey.Name);
+Console.WriteLine(updatedKey.Version);
+Console.WriteLine(updatedKey.Updated);
+```
+
+### Delete a Key
+`DeleteKey` deletes a key previously stored in the Key Vault. When [soft-delete][soft_delete] is not enabled for the Key Vault, this operation permanently deletes the key.
+
+```c#
+DeletedKey key = client.DeleteKey("key-name");
+
+Console.WriteLine(key.Name);
+Console.WriteLine(key.DeletedDate);
+```
+
+### List Keys
+This example lists all the keys in the specified Key Vault.
+
+```c#
+List<Response<KeyBase>> allKeys = Client.GetKeys();
+
```

Comment:
This line looks suspect.  Is it `List<Response<KeyBase>>` or `Response<List<KeyBase>>`?  And should `Client` be lowercase?

Reply:
GetKeys returns `IAsyncEnumerable<Response<KeyBase>>`

Reply:
Ah, since that's the sync API, should it be regular `IEnumerable` as well?

Reply:
yeah you are right, it was `IEnumerable<Response<KeyBase>>` and yeah, that is better. I'll change it